server/error_code/error_code.go

```diff
@@ -0,0 +1,72 @@
+// Package errcode is designed to create standardized API error codes.
+// The goal is that clients can reliably check against immutable error codes
+//
+// Two approaches can be takend
+// 1) centralized: define all errors in one module
+// 2) modular: define errors where they occur
+//
+// The centralized approach helps organize information.
+// All errors can be found in one file, and changes can be carefully reviewed.
+// The downside of centralized is the potential need to import types from other packages.
+//
+// A RegisteredCode should never be modified once committed (and released)
+package errcode
+
+import (
+	"fmt"
+	"net/http"
+)
+
+// RegisteredCode helps document that we are using a registered error code that must never change
+type RegisteredCode string
+
+const (
+	// InternalErrorCode means the operation placed the system is in an inconsistent or unrecoverable state
+	// Essentially a handled panic.
+	// This is the same as a HTTP 500, so it is not necessary to send this code when using HTTP.k
+	InternalErrorCode RegisteredCode = "internal"
+)
+
+// ErrorCode defines constant code functions Code() and HTTPCode().
+// Code returns a RegisteredCode defined in this module.
+// Most implementations of HTTPCode() will return the DefaultHTTPCode.
+// The Error() function is not constant: it converts the underlying struct data into a detailed string message.
+// The underlying struct data will also be returned as JSON, see ErrorCodeJSON.
+type ErrorCode interface {
+	Error() string // The Error interface
+	HTTPCode() int
+	Code() RegisteredCode
+}
```

Comment:
em, it seems a bit cumbersome to define different types for each kind of error and implement the corresponding methods in turn.
Can we provide some easy way to generate errors easily? For example, provide a function to generate directly through `{msg string, code RegisteredCode, httpCode int}`?

Reply:
That's a good idea to look at providing some additional convenience. We do need to be careful though to avoid encouraging just using a string message and try to make it easy to use structured data that can be useful to a client.